### PR TITLE
Ana lib fix and readme

### PR DESCRIPTION
**Pull request by siilieva**
Removing non-existing nestedfunctions c++ struct  - resulted in a warning during the build.
Add dependency to analysis_tools when building the analysis_cuts lib to prevent build issues on ubuntu. No issues were encountered on lxplus slc7/Alma9 machines even without the explicit dependency declaration. This proposed fix works on lxplus and ubuntu.

Update the README to include git pull, rebase and push instructions - a second review at these won't hurt.

### CODE REVIEW

File: analysis/cuts/CMakeLists.txt [hunk 1]
```diff
@@ -34,6 +34,6 @@ link_directories( ${LINK_DIRECTORIES})
 Set(HEADERS )
 Set(LINKDEF AnalysisCutsLinkDef.h)
 Set(LIBRARY_NAME snd_analysis_cuts)
-Set(DEPENDENCIES shipLHC veto)
+Set(DEPENDENCIES snd_analysis_tools shipLHC veto)
```
Comment (olantwin) [inline analysis/cuts/CMakeLists.txt:37]: I don't understand why this only popped up as an issue now. Or did we not test the original change?

File: README.md [hunk 2]
```diff
@@ -154,6 +156,81 @@ Commands are similar to the previous case, but without access to CVMFS you need
     alienv enter sndsw/latest
     ```
 
+# Update branch instructions
+
+To update one's local copy of a branch with the latest modifications of that branch on the upstream repository do:
+```bash
+cd sndsw
+git checkout <branch>
+git pull
```
Comment (olantwin) [inline README.md:165]: `git pull --rebase` would be better to prevent accidental merges.

File: README.md [hunk 3]
```diff
@@ -154,6 +156,81 @@ Commands are similar to the previous case, but without access to CVMFS you need
...
+cd ..
+source /cvmfs/sndlhc.cern.ch/SNDLHC-2023/Aug30/setUp.sh  # recommended latest version
+sndBuild
+```
+where `sndBuild` is a build command alias defined in the `setUp.sh` file.
+
+While working on your local copy of a feature `branch`, it is likely that the upstream `master` software changes (e.g. when a colleague merges their branch to the `master`).
```
Comment (olantwin) [inline README.md:172]: Maybe better: `e.g. because a pull request was merged into master`.

File: README.md [hunk 4]
```diff
@@ -154,6 +156,81 @@ Commands are similar to the previous case, but without access to CVMFS you need
...
+Then one propagates the updates of the upstream `master` to the local `branch` (`master -> branch`) doing:
+```bash
+cd sndsw
+git checkout master
+git pull
```
Comment (olantwin) [inline README.md:177]: Again: `git pull --rebase`

File: README.md [hunk 5]
```diff
@@ -154,6 +156,81 @@ Commands are similar to the previous case, but without access to CVMFS you need
...
+git checkout <branch>
+git pull
```
Comment (olantwin) [inline README.md:179]: This is only necessary in case of a branch with a remote.

File: README.md [hunk 6]
```diff
@@ -154,6 +156,81 @@ Commands are similar to the previous case, but without access to CVMFS you need
...
+git rebase master
```
Comment (olantwin) [inline README.md:180]: I'd generally recommend `git rebase -i master` (or `origin/master` in case you want to avoid updating your local master branch).

File: README.md [hunk 7]
```diff
@@ -154,6 +156,81 @@ Commands are similar to the previous case, but without access to CVMFS you need
...
+cd ..
+```
+This will put all commits of the feature branch on top of the `master` commits. It means the commit order won't be linear in time.
```
Comment (olantwin) [inline README.md:183]: I'm not sure whether the second sentence really help. I think it will just confuse people.

File: README.md [hunk 8]
```diff
@@ -154,6 +156,81 @@ Commands are similar to the previous case, but without access to CVMFS you need
...
+Note that before the `rebase` the local feature `branch` was updated via `git pull` too in case it changed on the upstream repository.
```
Comment (olantwin) [inline README.md:184]: `git pull --rebase`

File: README.md [hunk 9]
```diff
@@ -154,6 +156,81 @@ Commands are similar to the previous case, but without access to CVMFS you need
...
+
+If the `master` and the local `branch` both change the same line(s) in a file(s), a `conflict` rises up. Don't panic!
```
Comment (olantwin) [inline README.md:186]: arise/occur. Not rise up.

File: README.md [hunk 10]
```diff
@@ -154,6 +156,81 @@ Commands are similar to the previous case, but without access to CVMFS you need
...
+
+After solving any encountered `conflict`, build the updated branch doing:
+```bash
+source /cvmfs/sndlhc.cern.ch/SNDLHC-2023/Aug30/setUp.sh  # recommended latest version
+sndBuild
+```
+
+Following a `rebase` and after solving `conflicts`, to push one's local changes to the feature `branch` on the upstream repository do:
+```bash
+git push -f --set-upstream origin <branch>
```
Comment (olantwin) [inline README.md:198]: We should probably recommend using `--force-with-lease` instead.

File: README.md [hunk 11]
```diff
@@ -10,6 +10,8 @@ sndsw
 - [Build instructions](#build-instructions)
     - [On lxplus or systems with CVMFS](#on-lxplus-or-systems-with-cvmfs)
     - [On systems without access to CVMFS](#on-systems-without-access-to-cvmfs)
+- [Update branch instructions](#update-branch-instructions)
+- [Instructions to push changes to the upstream master](#instructions-to-push-changes-to-the-upstream-master)
```
Comment (olantwin) [inline README.md:14]: I think I'd change the structure a bit.

Maybe these should both the subsections to Contibuting code or we could make a section `Development`?

File: README.md [hunk 12]
```diff
@@ -154,6 +156,81 @@ Commands are similar to the previous case, but without access to CVMFS you need
...
+```
+Note the usage of `-f` (`--force`) flag in the push command, needed since the commit history of the feature `branch` is being re-written.
+
+# Instructions to push changes to the upstream master
+
+All changes of the `master` branch on the upstream repository go through a review process by the `sndsw` `core-developers`.
+To contribute to the `sndsw master`, follow the guidelines below:
+
+1. Prepare you feature `branch` and commit all your updates/fixes.
+2. Rebase the master on your branch:
```
Comment (olantwin) [inline README.md:208]: Rebase your branch on master!

File: README.md [hunk 13]
```diff
@@ -154,6 +156,81 @@ Commands are similar to the previous case, but without access to CVMFS you need
...
+    ```bash
+    cd sndsw
+    git checkout master
+    git pull
+    git checkout <branch>
+    git pull # for safety, perhaps not really needed here
+    git rebase master
+    cd ..
+    ```
```
Comment (olantwin) [inline README.md:217]: We should reference the update your branch section here instead of duplicating the instructions.

File: README.md [hunk 14]
```diff
@@ -154,6 +156,81 @@ Commands are similar to the previous case, but without access to CVMFS you need
...
+3. Resolve any `conflicts`. Do not create additional commits in doing so!
+   Verify this is the case reading the `git log` or inspecting via the graphical tool typing `gitk&`.
```
Comment (olantwin) [inline README.md:219]: spurious &?

File: README.md (hunk 13 above)
Comment (olantwin) [reply README.md:217]: Maybe "Update the branch and take care of potential conflicts, as described ..." and remove 2-4?

File: README.md [hunk 15]
```diff
@@ -154,6 +156,81 @@ Commands are similar to the previous case, but without access to CVMFS you need
...
+    git push -f --set-upstream origin <branch>
+    ```
+    Now the branch is ready for a merge.
+5. Go online to [the sndsw Pull request webpage](https://github.com/SND-LHC/sndsw/pulls) and click on `New pull request`.
+   Follow through the fields and verify the request goes from `branch` to `SND-LHC/sndsw`, i.e.
+   `base:master <- compare:branch`
+6. Leave a message explaining your changes, the need for them, etc.
+7. On the right, one can request review from a given member of the SND@LHC `core-developers`, e.g. when certain changes were already discussed with that person.
+   Otherwise, choose `core-developers`.
+8. Create the `Pull request`.
```
Comment (olantwin) [inline README.md:232]: If the changes are not yet ready for being merged, it is also possible to create a `draft pull request`.

For this, and many other topics, it would be good to link to the Github or git documentation.
We don't need to explain everything here.

**Review by olantwin — CHANGES_REQUESTED** [review]
I'd restructure and rephrase some of the README changes.

The other changes are fine, but I'm still a bit surprised that the CMake issue wasn't discovered earlier.

**Comment by olantwin** [comment]
At a higher level, we should maybe consider how much detail for git is appropriate for the README.

I think I'd lean towards explaining what to do, and relegating how to do it to other locations as much as possible (e.g. sndsw wiki, twiki, software tutorial, Github and Git documentation.

E.g.

> To contribute code, please make pull requests. In order to make the review as easy and quick  as possible, please make sure that your branch has been rebased against master.
> For detailed instructions, please see bla. For a short introduction to git and recommended git config options, please see here.

File: analysis/cuts/CMakeLists.txt (hunk 1 above)
Comment (siilieva) [reply analysis/cuts/CMakeLists.txt:37]: It is indeed peculiar this issue didn't appear before. All code was tested on scl7/rhel9 (lxplus machine OS) before the pull request  back when the original change happened.

File: README.md (hunk 12 above)
Comment (siilieva) [reply README.md:208]: oh yeah!

**Comment by siilieva** [comment]
> At a higher level, we should maybe consider how much detail for git is appropriate for the README.
> 
> I think I'd lean towards explaining what to do, and relegating how to do it to other locations as much as possible (e.g. sndsw wiki, twiki, software tutorial, Github and Git documentation.
> 
> E.g.
> 
> > To contribute code, please make pull requests. In order to make the review as easy and quick  as possible, please make sure that your branch has been rebased against master.
> > For detailed instructions, please see bla. For a short introduction to git and recommended git config options, please see here.


Once a Twiki/github etc page like that is set up it would be easy to cut down the Instructions.

File: README.md [hunk 16]
```diff
@@ -154,6 +157,67 @@ Commands are similar to the previous case, but without access to CVMFS you need
...
+```bash
+source /cvmfs/sndlhc.cern.ch/SNDLHC-2023/Aug30/setUp.sh  # recommended latest version
+sndBuild
+```
+
+Following a `rebase` and after solving `conflicts`, to push one's local changes to the feature `branch` on the upstream repository do:
+```bash
+git push --force-with-lease --set-upstream origin <branch>
+```
+Note the usage of `--force-with-lease` flag in the push command, which is needed since the commit history of the feature `branch` is being re-written and one makes sure the remote didn't change since the `git pull --rebase`.
```
Comment (olantwin) [inline README.md:204]: We could add a note here (or maybe just add a link to https://stackoverflow.com/a/52823955) to allow people to read up on `--force-with-lease`, as it's a lesser known flag, and people might wonder why it should be used instead of `--force`.

File: README.md [hunk 17]
```diff
@@ -154,6 +157,67 @@ Commands are similar to the previous case, but without access to CVMFS you need
...
+
+## Instructions to push changes to the upstream master
+
+All changes of the `master` branch on the upstream repository go through a review process by the `sndsw` `core-developers`.
+To contribute to the `sndsw master`, follow the guidelines below:
+
+1. Prepare you feature `branch` and commit all your updates/fixes.
+2. Update the branch and take care of potential conflicts, as described [above](#update-branch-instructions)
+3. Go online to [the sndsw Pull request webpage](https://github.com/SND-LHC/sndsw/pulls) and click on `New pull request`.
+   Follow through the fields and verify the request goes from `branch` to `SND-LHC/sndsw`, i.e.
```
Comment (olantwin) [inline README.md:214]: Maybe change "to the desired branch of SND-LHC/sndsw, i.e. base:target_branch", to cover the case where people want to contribute to AdvSND, future test beam branches whatever...

File: README.md [hunk 18]
```diff
@@ -154,6 +157,67 @@ Commands are similar to the previous case, but without access to CVMFS you need
...
+   `base:master <- compare:branch`
+4. Leave a message explaining your changes, the need for them, etc.
+5. On the right, one can request review from a given member of the SND@LHC `core-developers`, e.g. when certain changes were already discussed with that person.
```
Comment (olantwin) [inline README.md:217]: I would rephrase this, as it's a bit confusingly worded:
"In the right panel, under the 'Reviewers' heading, you can choose who should review the code, e.g. the expert for a sub-system or topic, or someone who you have already discussed the changes with."
The note about the core-developers isn't really necessary, as the branch requires a review from one of them anyway.

File: README.md [hunk 19]
```diff
@@ -154,6 +157,67 @@ Commands are similar to the previous case, but without access to CVMFS you need
...
+
+Following a `rebase` and after solving `conflicts`, to push one's local changes to the feature `branch` on the upstream repository do:
+```bash
+git push --force-with-lease --set-upstream origin <branch>
+```
+Note the usage of `--force-with-lease` flag in the push command, which is needed since the commit history of the feature `branch` is being re-written and one makes sure the remote didn't change since the `git pull --rebase`. 
+
+## Instructions to push changes to the upstream master
+
+All changes of the `master` branch on the upstream repository go through a review process by the `sndsw` `core-developers`.
```
Comment (olantwin) [inline README.md:208]: Maybe we should link the team here (https://github.com/orgs/SND-LHC/teams/core-developers), so that people can see who they are.

**Review by olantwin — CHANGES_REQUESTED** [review]
Much better!

Still a few nitpicks, but I think we're converging.

Happy to have the detail for now, if we are both in agreement that we should try to move any unnecessary detail elsewhere in the longer term.

File: README.md [hunk 20]
```diff
@@ -154,6 +157,67 @@ Commands are similar to the previous case, but without access to CVMFS you need
...
+cd sndsw
+git checkout <branch>
+git pull --rebase
+cd ..
+source /cvmfs/sndlhc.cern.ch/SNDLHC-2023/Aug30/setUp.sh  # recommended latest version
+sndBuild
+```
+where `sndBuild` is a build command alias defined in the `setUp.sh` file.
+
+While working on your local copy of a feature `branch`, it is likely that the upstream `master` software changes (e.g. because a pull request was merged into master).
```
Comment (olantwin) [inline README.md:175]: likely → possible

File: README.md [hunk 21]
```diff
@@ -10,6 +10,9 @@ sndsw
 - [Build instructions](#build-instructions)
     - [On lxplus or systems with CVMFS](#on-lxplus-or-systems-with-cvmfs)
     - [On systems without access to CVMFS](#on-systems-without-access-to-cvmfs)
+- [Development](#development)
```
[the remark this answers is not in the document]
Comment (siilieva) [reply README.md:13]: weren't we trying to reduce the text in the readme? why add an introduction if the (sub)section titles are clear?
I like your proposed titles "How to keep branches up to date" and "How to contribute code".
However, I wont swap the sections since with  instructions it is typically that one references backward and not forward, unlike papers, theses etc.

[the remark this answers is not in the document]
Comment (olantwin) [reply README.md:13]: > weren't we trying to reduce the text in the readme? why add an introduction if the (sub)section titles are clear?

I meant convert the existing section "Contributing code" into the introduction to "Development", so just moving exiting text around.

> However, I wont swap the sections since with instructions it is typically that one references backward and not forward, unlike papers, theses etc.

In that case, it might be good to briefly mention that development happens using pull requests and that this is easiest to do if everyone keeps their branches updated. Then the order makes sense.

File: README.md (hunk 16 above)
Comment (siilieva) [reply README.md:204]: I aimed at explaining the need for `--force-and-lease` with the text, but I will simply add the suggested link.

Comment (olantwin) [reply README.md:204]: You're right, I didn't notice the last sub-clause.
May I suggest adding in parentheses after the `--rebase`: (this differentiates [`--force-with-lease`](link goes here) from the more common `--force`)
or something like that.

File: README.md (hunk 17 above)
Comment (siilieva) [reply README.md:214]: We diverged here from the original idea.
There is the separation of topics : update branch and push to a feature branch (testbeam AdvSND are feature branches)
and then comes  push to master

File: README.md (hunk 18 above)
Comment (siilieva) [reply README.md:217]: ok for the first half.
>>The note about the core-developers isn't really necessary, as the branch requires a review from one of them anyway.

User don't know that beforehand.

Comment (olantwin) [reply README.md]: > User don't know that beforehand.

But do they need to? They won't be able to merge anyway by themselves.

File: README.md (hunk 17 above)
Comment (olantwin) [reply README.md]: > There is the separation of topics : update branch and push to a feature branch (testbeam AdvSND are feature branches) and then comes push to master

No, there isn't. There is no difference between making a PR to master or to longer-term branches like AdvSND where multiple people might work together, and the instructions are the same (apart from the review requirement due to different branch protection rules, which we don't need to mention).

Comment (siilieva) [reply README.md]: The README so far doesn't aim to cover the case when a feature branch isn't checked out from the master. Here I gave one scenario - contribute to master.
However, it takes indeed a bit re-organizing to add a more generic case, so let me try

Comment (olantwin) [reply README.md]: Thanks!

I don't think much reorganisation is needed:
In this point, you can just add in parentheses behind `compare:branch` "(or substitute the target branch for master, if you want to contribute to another branch)."
I don't think any other changes are needed.
Maybe in the updating instructions, one could replace master with "base branch (usually master)".

Comment (siilieva) [reply README.md]: yes, indeed it was a few changes. And it is much better now!

File: README.md [hunk 22]
```diff
@@ -245,6 +248,67 @@ Two methods implemented, hitMaps(Nev = -1) and eventTime().
...
+or the [Software and Analysis Mattermost channel](https://mattermost.web.cern.ch/sndlhc-daq/channels/software-and-analysis).
+One can verify the `commit history` reading the `git log` or inspecting via the graphical tool typing `gitk`.
+
+After solving any encountered `conflict`, build the updated branch as per the [Build instructions](#build-instructions).
+
+Following a `rebase` and after solving `conflicts`, to push one's local changes to the feature `branch` on the upstream repository do:
+```bash
+git push --force-with-lease --set-upstream origin <branch>
+```
+Follow the link for further reading why we use [`--force-with-lease` flag](https://stackoverflow.com/questions/52823692/git-push-force-with-lease-vs-force/52823955#52823955).
```
Comment (olantwin) [inline README.md:293]: use → recommend

**Review by olantwin — APPROVED** [review]
Tiny nitpicks, but generally this is now basically perfect!

Thanks for putting up with my comments!

**Comment by olantwin** [comment]
One comment maybe: generally for discussing changes it would be good to make individual commits. They can then later all be squashed into one in a final rebase before merging.

I had trouble following all the small changes between the different versions of this commit.

**Comment by siilieva** [comment]
> One comment maybe: generally for discussing changes it would be good to make individual commits. They can then later all be squashed into one in a final rebase before merging.
> 
> I had trouble following all the small changes between the different versions of this commit.

true, noted! 
Thanks for your feedback!
Btw we deserve a treat for this effort!

**Comment by olantwin** [comment]
> > One comment maybe: generally for discussing changes it would be good to make individual commits. They can then later all be squashed into one in a final rebase before merging.
> > I had trouble following all the small changes between the different versions of this commit.
> 
> true, noted! Thanks for your feedback! Btw we deserve a treat for this effort!

Sure, treats next week at the CM?
